### ISSUE TITLE
mergeFrom for repeated fields

### ISSUE THREAD

**Issue by 1zg12** [protocolbuffers/protobuf#8056]
**What language does this apply to?**
Java

If it's a proto syntax change, is it for proto2 or proto3?
proto3

If it's about generated code change, what programming language?
Java

**Describe the problem you are trying to solve.**
In the current `mergeFrom` implementation, it will override existing fields if those are singular. However, for repeated, `mergeFrom` would simply concatenate the fields.

This will cause a lot of duplicates.

As an example, let's say we have a wrapper (`orders.proto`) that has 10 (`order.proto`) at the moment.
With new messages coming in (which contains 5 new orders and 5 existing orders), the mergeFrom as the name suggested, I expect it would only concatenate new orders which would result in 15 orders.

**Describe the solution you'd like**

Instead of using `addAll` method from `java.util.Collections.emptyList()`, we can either
> use `set` to drop the duplicate

> iterate through the new list, and only if add the order not already in existance

**Describe alternatives you've considered**

**Additional context**

We could add an overloaded method of `mergeFrom` to give this option to mergeWithoutDuplicte.
```
public Builder mergeFrom(... other, boolean mergeWithoutDuplicte) {

  .....
 if(mergeWithoutDuplicte){
   //drop the duplicate
 }
}
```

**Comment by ennerf** [comment]
I'm not a Protobuf dev, but I think this needs a bit more clarification. Are you suggesting that merging repeated (what about composite?) fields should behave like singular values and get overwritten with the values in the other message, or are you suggesting some sort of smart merge based on the values of the data?

Protobuf is a serialization format that has no context about what the values represent, so it'd be the wrong place for implementing some sort of smart value filtering. (Why are you sending unwanted duplicate orders in the first place?)

**Comment by Th4n** [comment]
I was wondering what's the state of this issue? API wise I'd say that merge from should accept some strategy to specify whether repeated fields should be replaced or merged. As for the use case you can imagine bidirectional streaming where the server sends full tick only once, then send small delta updates for fields that have changed. As the client only wants to know the current state it's convenient to mergeFrom proto messages in order to get one object with all the fields containing the most recent value.

**Comment by fowles** [comment]
Changing the existing semantics for this is definitely not something we want to pursue.  There is too much code that would break.  Adding a policy object is possible, but is a far larger piece of design and change then we will likely take on in the near term

**Comment by Th4n** [comment]
I fully understand the hesitation with this feature, and agree that's not exactly in the scope of protobufs. However, it seems like this request is popping here and there since 2012 so maybe it's worth reconsidering it?

In case I'd like to fork the project and compile my own version, could you provide some more insights regarding what can go wrong? I don't have much knowledge about protobuf internals, however, since most of the work for merge is already done, and list case with replacing value should (hopefully?) be easier than merging itself I was wondering what are the possible places I could mess something up? Would replacing `ListFieldSchema` in `ManifestSchemaFactory (based on java implementation) with my own version doing replace rather than merge ` be a good place to start looking?

**Comment by ennerf** [comment]
**Specification**

I assume that one of the reasons this wasn't implemented is because the [specification ](https://developers.google.com/protocol-buffers/docs/encoding#packed) states that repeated fields do not have to be serialized as a contiguous block

> If a proto2 message definition has repeated elements (without the [packed=true] option), the encoded message has zero or more key-value pairs with the same field number. These repeated values do not have to appear consecutively; they may be interleaved with other fields. The order of the elements with respect to each other is preserved when parsing, though the ordering with respect to other fields is lost. In proto3, repeated fields use packed encoding, which you can read about below.

This means that if there is a repeated field `repeated Type list = 1`, messages serialized as `1,1,1,2,3` are equivalent to `1,2,1,3,1`. Implementing a merge-replace would cause issues with the second case. I don't think that there are any current implementations that serialize messages like this, but there are probably many users that rely on this mechanism for appending data to existing messages.

The `packed=true` tag results in a more efficient encoding `[key][length][<contiguous-list-of-all-values>]` for scalar types, but my my understanding is that packed fields can also be split into multiple packed blocks.

**Brainstorming**

It could be interesting to add an option for the merge behavior, e.g., `repeated Type field = 1 [on_merge = APPEND || REPLACE];` where REPLACE would clear and replace the list with new content, i.e., similar behavior to merging scalars.

Protoc 3.17.3 generates the following parsing code for `repeated NestedMessage repeated_nested_message  = 48;`

```java
// parsing code generated for a repeated message
case 386: {
  if (!((mutable_bitField1_ & 0x00000100) != 0)) {
    repeatedNestedMessage_ = new java.util.ArrayList<protos.test.protobuf.TestAllTypes.NestedMessage>();
    mutable_bitField1_ |= 0x00000100;
  }
  repeatedNestedMessage_.add(
      input.readMessage(protos.test.protobuf.TestAllTypes.NestedMessage.PARSER, extensionRegistry));
  break;
}
```

It looks like the switch statement is executed every time a tag is encountered, so you'd have to add something to keep track of whether you are inside a continuous block of repeated values. The easiest way I can think of would be to check whether the tag has changed since the last iteration, e.g., something like

```java
// untested example
boolean done = false;
int previousTag = 0;
while (!done) {
  int tag = input.readTag();
  switch (tag) {
    // ... other cases
    case 386: {
    if (!((mutable_bitField1_ & 0x00000100) != 0)) {
      repeatedNestedMessage_ = new java.util.ArrayList<protos.test.protobuf.TestAllTypes.NestedMessage>();
      mutable_bitField1_ |= 0x00000100;
    }
    if (previousTag != tag) { // generate if on_merge == REPLACE
      list.clear();
    }
    repeatedNestedMessage_.add(
        input.readMessage(protos.test.protobuf.TestAllTypes.NestedMessage.PARSER, extensionRegistry));
    break;
  }
  // ... other cases
  }
  previousTag = tag;
}
```

It has been a while since I looked at the protoc sources, but as far as I remember you would need to take a look at the case statement generation in [ImmutableMessageGenerator::GenerateParsingConstructor](https://github.com/protocolbuffers/protobuf/blob/master/src/google/protobuf/compiler/java/java_message.cc#L1229-L1235) and the `Repeated<xxx>Field::GenerateParsingCode` implementations, e.g., [RepeatedImmutableMessageFieldGenerator::GenerateParsingCode](https://github.com/protocolbuffers/protobuf/blob/master/src/google/protobuf/compiler/java/java_message_field.cc#L1326-L1345) for repeated message fields (this part is a bit of a mess for code duplication).
